File: languages/objective-c/tests/PositionTapeTests.m
#import <Foundation/Foundation.h>
#import "../src/PositionTape.h"

static void PTAssert(BOOL condition, NSString *message) {
    if (!condition) {
        [NSException raise:@"PositionTapeTestFailure" format:@"%@", message];
    }
}

int main(void) {
    @autoreleasepool {
        PTAssert([[PositionTape Generate:0] isEqualToString:@""], @"zero length");
        PTAssert([[PositionTape Generate:11] isEqualToString:@"12345678911"], @"basic generation");
        PTAssert([PositionTape Generate:100].length == 100, @"exact length");
        PTAssert([PositionTape GenerateMarkerComplete:100].length == 101, @"marker complete 100");
        PTAssert([PositionTape GenerateMarkerComplete:10000].length == 10003, @"marker complete 10000");

        PTValidationResult *valid = [PositionTape Validate:[PositionTape Generate:250] expectedLength:250];
        PTAssert(valid.isValid, @"valid tape");

        PTValidationResult *truncated = [PositionTape Validate:[PositionTape Generate:40] expectedLength:50];
        PTAssert(!truncated.isValid, @"truncated invalid");
        PTAssert([truncated.truncationPoint integerValue] == 41, @"truncation point");

        PTMismatch *mismatch = [PositionTape FindFirstMismatch:[PositionTape Generate:20]
                                                      received:[[PositionTape Generate:19] stringByAppendingString:@"X"]];
        PTAssert(mismatch != nil && mismatch.position == 20, @"mismatch");

        NSLog(@"OK objective-c");
    }
    return 0;
}

File: languages/delphi/tests/position_tape_tests.pas
program PositionTapeTests;

{$mode objfpc}
{$H+}

uses
  SysUtils,
  PositionTape;

procedure Check(Condition: Boolean; const MessageText: string);
begin
  if not Condition then
    raise Exception.Create(MessageText);
end;

var
  Validation: TValidationResult;
  Mismatch: TMismatch;
begin
  Check(Generate(0) = '', 'zero length');
  Check(Generate(11) = '12345678911', 'basic generation');
  Check(Length(Generate(100)) = 100, 'exact length');
  Check(Length(GenerateMarkerComplete(100)) = 101, 'marker complete 100');
  Check(Length(GenerateMarkerComplete(10000)) = 10003, 'marker complete 10000');

  Validation := Validate(Generate(250), 250);
  Check(Validation.IsValid, 'valid tape');

  Validation := Validate(Generate(40), 50);
  Check(not Validation.IsValid, 'truncated invalid');
  Check(Validation.HasTruncationPoint and (Validation.TruncationPoint = 41), 'truncation point');

  Mismatch := FindFirstMismatch(Generate(20), Generate(19) + 'X');
  Check(Mismatch.HasValue and (Mismatch.Position = 20), 'mismatch');

  WriteLn('OK delphi');
end.
